### PR TITLE
Improving mfa tests coverage and format

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Improving mfa tests coverage and format.

This speeds up the tests, because is not necessary at all to instantiate the mainhelper to test this service, and at the same time raises the coverage of this service to nearly the 100% (This is a super tiny service, so this is not super impressive, but well... this is one piece of code that now is properly tested).

#### Ticket
[MM-22939](https://mattermost.atlassian.net/browse/MM-22939)

### CODE REVIEW

File: services/mfa/mfa_test.go [hunk 1]
```diff
@@ -27,26 +28,45 @@ func TestGenerateSecret(t *testing.T) {
...
-	storeMock.On("User").Return(&userStoreMock)
 
-	mfa := Mfa{configService, &storeMock}
+	t.Run("fail on disabled mfa", func(t *testing.T) {
+		wrongConfig := model.Config{}
+		wrongConfig.SetDefaults()
+		wrongConfig.ServiceSettings.EnableMultifactorAuthentication = model.NewBool(false)
+		wrongConfigService := testutils.StaticConfigService{Cfg: &wrongConfig}
+		mfa := New(wrongConfigService, nil)
+		_, _, err := mfa.GenerateSecret(user)
```
Comment (agnivade) [inline services/mfa/mfa_test.go]: This is an AppError with a specific id: `mfa.mfa_disabled.app_error`. Let's check that to be more specific of the error returned.

File: services/mfa/mfa_test.go [hunk 2]
```diff
@@ -27,26 +28,45 @@ func TestGenerateSecret(t *testing.T) {
...
+		userStoreMock := mocks.UserStore{}
+		userStoreMock.On("UpdateMfaSecret", user.Id, mock.AnythingOfType("string")).Return(func(userId string, secret string) *model.AppError {
+			return model.NewAppError("GenerateQrCode", "mfa.generate_qr_code.save_secret.app_error", nil, "", http.StatusInternalServerError)
+		})
+		storeMock.On("User").Return(&userStoreMock)
 
-	assert.Len(t, secret, 32)
+		mfa := New(configService, &storeMock)
+		_, _, err := mfa.GenerateSecret(user)
+		require.NotNil(t, err)
```
Comment (agnivade) [inline services/mfa/mfa_test.go]: Same as above.

File: services/mfa/mfa_test.go [hunk 3]
```diff
@@ -27,26 +28,45 @@ func TestGenerateSecret(t *testing.T) {
...
+	})
 
-	require.NotEmpty(t, img, "no image set")
+	t.Run("Successful generate secret", func(t *testing.T) {
```
Comment (agnivade) [inline services/mfa/mfa_test.go]: Is it also possible to test for `qr.Encode(authLink, qr.H)` by passing some garbage `SiteURL` ?

File: services/mfa/mfa_test.go [hunk 4]
```diff
@@ -27,26 +28,45 @@ func TestGenerateSecret(t *testing.T) {
...
 
-	config.ServiceSettings.EnableMultifactorAuthentication = model.NewBool(false)
+		mfa := New(configService, &storeMock)
 
-	_, _, err = mfa.GenerateSecret(user)
-	require.NotNil(t, err)
+		secret, img, err := mfa.GenerateSecret(user)
+		require.Nil(t, err)
+		assert.Len(t, secret, 32)
+		require.NotEmpty(t, img, "no image set")
```
Comment (agnivade) [inline services/mfa/mfa_test.go]: Is it also possible to exactly verify the image returned ?

File: services/mfa/mfa_test.go [hunk 5]
```diff
@@ -72,24 +92,168 @@ func TestGetIssuerFromUrl(t *testing.T) {
...
-		return nil
+
+	t.Run("fail on disabled mfa", func(t *testing.T) {
+		wrongConfig := model.Config{}
+		wrongConfig.SetDefaults()
+		wrongConfig.ServiceSettings.EnableMultifactorAuthentication = model.NewBool(false)
+		wrongConfigService := testutils.StaticConfigService{Cfg: &wrongConfig}
+		mfa := New(wrongConfigService, nil)
+		err := mfa.Activate(user, "not-important")
+		require.NotNil(t, err)
```
Comment (agnivade) [inline services/mfa/mfa_test.go]: Same as above.

File: services/mfa/mfa_test.go [hunk 6]
```diff
@@ -72,24 +92,168 @@ func TestGetIssuerFromUrl(t *testing.T) {
...
 	})
-	storeMock.On("User").Return(&userStoreMock)
 
-	mfa := Mfa{configService, &storeMock}
+	t.Run("fail on wrongly formatted token", func(t *testing.T) {
+		mfa := New(configService, nil)
+		err := mfa.Activate(user, "invalid-token")
+		require.NotNil(t, err)
```
Comment (agnivade) [inline services/mfa/mfa_test.go]: Does this test differentiate from the other one in the error returned ? Specifically I would want to test for `mfa.activate.authenticate.app_error` and `mfa.activate.bad_token.app_error`.

File: services/mfa/mfa_test.go [hunk 7]
```diff
@@ -72,24 +92,168 @@ func TestGetIssuerFromUrl(t *testing.T) {
...
+		storeMock := mocks.Store{}
+		userStoreMock := mocks.UserStore{}
+		userStoreMock.On("UpdateMfaActive", user.Id, true).Return(func(userId string, active bool) *model.AppError {
+			return model.NewAppError("Activate", "mfa.activate.save_active.app_error", nil, "", http.StatusInternalServerError)
+		})
+		storeMock.On("User").Return(&userStoreMock)
+
+		mfa := New(configService, &storeMock)
+		err := mfa.Activate(user, fmt.Sprintf("%06d", token))
+		require.NotNil(t, err)
```
Comment (agnivade) [inline services/mfa/mfa_test.go]: Same as above.

File: services/mfa/mfa_test.go (hunk 3 above)
Comment (jespino) [reply services/mfa/mfa_test.go]: It is not that easy I think, I tried but the function `Encode` is pretty robust by itself.

File: services/mfa/mfa_test.go (hunk 4 above)
Comment (jespino) [reply services/mfa/mfa_test.go]: I don't think so, the image returned is generated based on the data passed to the library, we could use a qr reader library to read the qr on the image en verify if this match properly, but that is testing the third party library, no our code.

**Pull request merged**